Adding S3 nuke support

saurabh-hirani opened the pull request:
This PR adds support to delete S3 buckets as requested in https://github.com/gruntwork-io/cloud-nuke/issues/58 

It ensures that the AWS S3 list bucket operation - which lists all buckets irrespective of region - is not called multiple times for each region and uses cached entries as highlighted in the above issue discussion. 

Deletion also uses ```WaitUntilBucketNotExists``` to ensure bucket is deleted. It also supports edge cases where in bucket is deleted by some other means after listing but before deletion.

**Demo:**

1. Added s3 to list of resources

```
go run main.go aws --list-resource-types | grep s3

s3
```

2. Run S3 related tests:

```
go test -run S3 -v ./...

=== RUN   TestListS3Buckets
time="2020-02-29T22:52:35+08:00" level=info msg="Random region chosen: ap-northeast-1"
time="2020-02-29T22:52:42+08:00" level=info msg="Awaiting creation - Bucket: cloud-nuke-test-gnd7gn"
time="2020-02-29T22:52:47+08:00" level=info msg="Deleting - 1 S3 Buckets in region ap-northeast-1"
time="2020-02-29T22:52:47+08:00" level=info msg="Deleting - 1/1 - S3 Bucket: cloud-nuke-test-gnd7gn"
time="2020-02-29T22:52:48+08:00" level=info msg="Awaiting deletion - 1/1 - Bucket: cloud-nuke-test-gnd7gn"
time="2020-02-29T22:52:48+08:00" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-gnd7gn - deleted"
--- PASS: TestListS3Buckets (13.35s)
```

TestListS3Buckets - creates, verifies S3 bucket existence, deletes and verifies S3 bucket non existence. 

```
=== RUN   TestNukeS3Buckets
time="2020-02-29T22:52:52+08:00" level=info msg="Random region chosen: ca-central-1"
time="2020-02-29T22:52:55+08:00" level=info msg="Awaiting creation - Bucket: cloud-nuke-test-npymlv"
time="2020-02-29T22:52:56+08:00" level=info msg="Deleting - 1 S3 Buckets in region ca-central-1"
time="2020-02-29T22:52:56+08:00" level=info msg="Deleting - 1/1 - S3 Bucket: cloud-nuke-test-npymlv"
time="2020-02-29T22:52:57+08:00" level=info msg="Awaiting deletion - 1/1 - Bucket: cloud-nuke-test-npymlv"
time="2020-02-29T22:52:57+08:00" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-npymlv - deleted"
time="2020-02-29T22:53:14+08:00" level=info msg="Deleting - 1 S3 Buckets in region ca-central-1"
time="2020-02-29T22:53:15+08:00" level=info msg="S3 Bucket cloud-nuke-test-npymlv exists check error code - NotFound"
time="2020-02-29T22:53:15+08:00" level=warning msg="Failed to find bucket - cloud-nuke-test-npymlv - ignoring as bucket might be deleted."
--- PASS: TestNukeS3Buckets (27.44s)
PASS
```

TestNukeS3Buckets - creates, deletes and verifies S3 bucket deletion - it also tries to delete the same bucket twice to try out the case of bucket present during listing but deleted by some other means before deletion through cloud-nuke.

3. Run S3 deletion with dry run:

```
go run main.go aws --resource-type s3 --region ap-northeast-1 --dry-run


INFO[2020-02-29T23:00:41+08:00] Retrieving active AWS resources in [ap-northeast-1]
INFO[2020-02-29T23:00:41+08:00] Checking region [1/1]: ap-northeast-1
INFO[2020-02-29T23:00:46+08:00] The following AWS resources are going to be nuked:
INFO[2020-02-29T23:00:46+08:00] * test-bucket-1-ap-northeast-1

INFO[2020-02-29T23:00:46+08:00] * test-bucket-2-ap-northeast-1

INFO[2020-02-29T23:00:46+08:00] Not taking any action as dry-run set to true.
```

4. Run S3 deletion without dry-run:

```

go run main.go aws --resource-type s3 --region ap-northeast-1

INFO[2020-02-29T23:00:41+08:00] Retrieving active AWS resources in [ap-northeast-1]
INFO[2020-02-29T23:00:41+08:00] Checking region [1/1]: ap-northeast-1
INFO[2020-02-29T23:00:46+08:00] The following AWS resources are going to be nuked:
INFO[2020-02-29T23:00:46+08:00] * s3-test-bucket-1-ap-northeast-1

INFO[2020-02-29T23:00:46+08:00] * s3-test-bucket-2-ap-northeast-1


THE NEXT STEPS ARE DESTRUCTIVE AND COMPLETELY IRREVERSIBLE, PROCEED WITH CAUTION!!!

Are you sure you want to nuke all listed resources? Enter 'nuke' to confirm (or exit with ^C): nuke

INFO[2020-02-29T23:01:06+08:00] Terminating 2 resources in batches
INFO[2020-02-29T23:01:06+08:00] Deleting - 2 S3 Buckets in region ap-northeast-1
INFO[2020-02-29T23:01:07+08:00] Deleting - 1/2 - S3 Bucket: test-bucket-1
INFO[2020-02-29T23:01:08+08:00] Awaiting deletion - 1/2 - Bucket: test-bucket-1
INFO[2020-02-29T23:01:08+08:00] [OK] - 1/2 - Bucket: test-bucket-1 - deleted
INFO[2020-02-29T23:01:08+08:00] Deleting - 2/2 - S3 Bucket: test-bucket-2
INFO[2020-02-29T23:01:09+08:00] Awaiting deletion - 2/2 - Bucket: test-bucket-2
INFO[2020-02-29T23:01:09+08:00] [OK] - 2/2 - Bucket: test-bucket-2 - deleted
INFO[2020-02-29T23:01:09+08:00] [OK] - 2/2 - S3 bucket(s) deleted in ap-northeast-1
```

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+		// Verify if that is the case.
+		exists, err := checkS3BucketExists(svc, *bucketName)
+		if err != nil {
+			return delCount, err
```
I think we'd want to attempt to delete all the buckets before returning the error.

saurabh-hirani replied:
Ok - will remove the checkS3BucketExists and ensure to delete max before erroring in case few buckets are removed before deletion.

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+		}
+
+		logging.Logger.Infof("[OK] - %d/%d - Bucket: %s - deleted", bucketCount, totalCount, *bucketName)
+		delCount++
```
This whole routine (deleting a single bucket) should be extracted out into a separate function for easier readability.

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+
+		iter := s3manager.NewDeleteListIterator(svc, &s3.ListObjectsInput{
+			Bucket: bucketName,
+		})
```
Does this work with versioning?

brikis98 replied:
And non empty buckets?

saurabh-hirani replied:
Checked with non-empty buckets. Didn't check with versioning. But will check and add test cases.

yorinasub17 commented on aws/s3_test.go:
```diff
@@ -0,0 +1,148 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func createTestS3Bucket(t *testing.T, session *session.Session, bucketName string) {
```
This is a good start for testing, but I'd like to see:

- A bucket with objects to make sure cloud-nuke handles that correctly
- A bucket with versioning enabled, and a deleted object version to make sure that is handled.

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+			return delCount, err
+		}
+
+		logging.Logger.Infof("[OK] - %d/%d - Bucket: %s - deleted", bucketCount, totalCount, *bucketName)
```
Nice logging 👍

brikis98 commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+			Bucket: bucketName,
+		})
+
+		if err := s3manager.NewBatchDeleteWithClient(svc).Delete(aws.BackgroundContext(), iter); err != nil {
```
By the name, it sounds like `BatchDelete` is used to delete multiple buckets, but it looks like we're using it to delete one at a time? Could we pass it the whole list, broken down into batches, instead?

yorinasub17 replied:
I was confused too, and then I looked up the docs and it turns out this API deletes all objects that match the filter passed in to the iterator, where the filter is `ListObjectsInput`: listing objects of a bucket.

Either way I think a comment before this section, or wrapping this in a function named `emptyBucket` would make it clearer.

saurabh-hirani replied:
Understood. Yes BatchDelete was to take care of non-empty buckets. Will call it nukeS3BucketObjects for consistentcy.

saurabh-hirani commented:
Thanks for the feedback @yorinasub17 and @brikis98. I will make the changes and update this thread.

saurabh-hirani commented on aws/aws.go:
```diff
@@ -393,6 +395,52 @@ func GetAllResources(targetRegions []string, excludeAfter time.Time, resourceTyp
+					return nil, errors.WithStackTrace(err)
+				}
+
+				if resourcesCache["S3"] == nil {
```
Checked. not required. Will fix.

saurabh-hirani commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+
+	input := &s3.ListBucketsInput{}
+
+	output, err := svc.ListBuckets(input)
```
Weirdly enough, ListBuckets does not paginate - https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.ListBuckets. I remember checking it earlier.

saurabh-hirani commented on aws/s3.go:
```diff
@@ -0,0 +1,159 @@
+		result, err := svc.GetBucketLocation(input)
+		if err != nil {
+			// It might be possible that between listing the bucket and finding the location - the target bucket got deleted
+			// Verify if that is the case.
```
Will check.

saurabh-hirani commented on aws/aws.go:
```diff
@@ -393,6 +395,52 @@ func GetAllResources(targetRegions []string, excludeAfter time.Time, resourceTyp
+			if len(bucketNamesPerRegion[region]) > 0 {
+				s3Buckets.Names = awsgo.StringValueSlice(bucketNames)
+				resourcesInRegion.Resources = append(resourcesInRegion.Resources, s3Buckets)
+			}
```
This was added because ```bucketNamesPerRegion``` contains only those regions which have buckets. When we are iterating through all AWS regions in the outer loop - we might encounter a region which has no bucket hence no entry in ```bucketNamesPerRegion``` - in that case ```bucketNamesPerRegion[region]``` will fail. But I will check it again and verify if that is the case.

saurabh-hirani commented:
@yorinasub17 @brikis98 - updated as per feedback with 2 items pending. Highlights:

Done:

1. Refactored ```s3_test.go``` to have separate tests for nuking empty bucket, bucket with items.
2. Refactored ```s3.go``` accordingly and cleaned up main nukeAllS3Buckets function to call ```nukeAllS3BucketObjects``` and ```nukeEmptyS3Bucket```
3. Updated ```aws.go``` as per nit-pick feedback. 

Pending:

1. Current code does not support deleting versioned bucket with objects. I couldn't find a good batch delete way for versioned buckets. Will probably detect bucket type and use right type of nuking.
2. Preventing deletion based on hard coded tag means adding metadata to the buckets returned by ```getAllS3Buckets``` to know bucket properties. As it is to find if bucket is versioned or not, we will have to add that metadata.

Will work on the pending items in the next few days and update.

saurabh-hirani commented:
Updated codebase as per the above feedback. Deleting S3 buckets has the following features:

1. Support to delete empty, non-empty and versioned buckets.
2. Ability to filter out buckets tagged with "cloud-nuke-excluded=true" flag. 
3. Batched deletion and deletion as per max AWS limits (1000 objects per deletion run). The batched deletion number is internal to the code and is not exposed to the end user.
4. Deleting buckets having specific substring. This feature is also internal to the code and is not exposed to the end user. It was added to make testing easier. 

All the above features have been verified via the following test cases. All the tests try to ensure deletion of the bucket in the end - hence there are some messages which are errors related to deleting duplicate buckets but they are not tied to test failure. 

1. List S3 buckets for deletion when no filter tag (cloud-nuke-excludeded=true) is present:

``` 
=== RUN   TestListEmptyS3BucketWithoutFilterTag
time="2020-03-10T16:23:31+05:30" level=info msg="Random region chosen: us-east-2"
time="2020-03-10T16:23:33+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"
time="2020-03-10T16:23:38+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:23:38+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - empty - skipping object deletion"
time="2020-03-10T16:23:39+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
--- PASS: TestListEmptyS3BucketWithoutFilterTag 
```

2. List S3 buckets for deletion when filter tag is present:

```
=== RUN   TestListEmptyS3BucketWithFilterTag
time="2020-03-10T16:23:41+05:30" level=info msg="Random region chosen: ap-northeast-1"
time="2020-03-10T16:23:42+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"
time="2020-03-10T16:23:46+05:30" level=info msg="Skipping - Bucket cloud-nuke-test-xxxx - matched tag filter"
time="2020-03-10T16:23:46+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:23:47+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - empty - skipping object deletion"
time="2020-03-10T16:23:48+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
--- PASS: TestListEmptyS3BucketWithFilterTag 
```

3. Deleting empty bucket:

```
=== RUN   TestNukeEmptyS3Bucket
time="2020-03-10T16:23:49+05:30" level=info msg="Random region chosen: us-east-2"
time="2020-03-10T16:23:49+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"
time="2020-03-10T16:23:52+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:23:53+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - empty - skipping object deletion"
time="2020-03-10T16:23:54+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:23:54+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:23:54+05:30" level=error msg="[Failed] - 1/1 - Bucket: cloud-nuke-test-xxxx - object deletion error - NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: xxxx, host id: xxxx"
--- PASS: TestNukeEmptyS3Bucket 
```

4. Deleting non-empty bucket - without any object versioning - deletion batch size (1000) > number of objects (10):

```
=== RUN   TestNukeS3BucketWithoutVersioningAllObjects
time="2020-03-10T16:23:59+05:30" level=info msg="Random region chosen: sa-east-1"
time="2020-03-10T16:23:59+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"
time="2020-03-10T16:24:03+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 0-0"
time="2020-03-10T16:24:03+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 0-1"
time="2020-03-10T16:24:03+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 0-2"
time="2020-03-10T16:24:04+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f3.txt - content: 0-3"
.
.
time="2020-03-10T16:24:06+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 0-9"
time="2020-03-10T16:24:06+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:24:08+05:30" level=info msg="Deleting - Bucket: cloud-nuke-test-xxxx - objects: 10"
time="2020-03-10T16:24:09+05:30" level=info msg="[OK] - 1-10 objects of batch 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:11+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:11+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:24:12+05:30" level=error msg="[Failed] - 1/1 - Bucket: cloud-nuke-test-xxxx - object deletion error - NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: xxxx, host id: xxxx"
--- PASS: TestNukeS3BucketWithoutVersioningAllObjects 
```

5. Deleting non-empty bucket - without any object versioning - deletion batch size (2) < number of objects (10):

```
=== RUN   TestNukeS3BucketWithoutVersioningBatchObjects
time="2020-03-10T16:24:14+05:30" level=info msg="Random region chosen: eu-west-1"
time="2020-03-10T16:24:14+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"
time="2020-03-10T16:24:17+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 0-0"
time="2020-03-10T16:24:17+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 0-1"
time="2020-03-10T16:24:18+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 0-2"
time="2020-03-10T16:24:18+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f3.txt - content: 0-3"
.
.
.
time="2020-03-10T16:24:21+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 0-9"
time="2020-03-10T16:24:21+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:24:23+05:30" level=info msg="Deleting - Bucket: cloud-nuke-test-xxxx - objects: 10"
time="2020-03-10T16:24:23+05:30" level=info msg="[OK] - 1-2 objects of batch 1/5 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:25+05:30" level=info msg="[OK] - 3-4 objects of batch 2/5 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:27+05:30" level=info msg="[OK] - 5-6 objects of batch 3/5 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:28+05:30" level=info msg="[OK] - 7-8 objects of batch 4/5 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:29+05:30" level=info msg="[OK] - 9-10 objects of batch 5/5 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:31+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:32+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:24:32+05:30" level=error msg="[Failed] - 1/1 - Bucket: cloud-nuke-test-xxxx - object deletion error - NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: xxxx, host id: xxxxx"
--- PASS: TestNukeS3BucketWithoutVersioningBatchObjects (20.60s)
```

6. Deleting non-empty bucket - with object versioning - deletion batch size (1000) > number of objects (30):

Versioned objects are created by add the same object multiple times with different content.

```
=== RUN   TestNukeS3BucketWithVersioningAllObjects
time="2020-03-10T16:24:33+05:30" level=info msg="Random region chosen: ca-central-1"
time="2020-03-10T16:24:33+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"
time="2020-03-10T16:24:37+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 0-0"
time="2020-03-10T16:24:38+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 0-1"
time="2020-03-10T16:24:38+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 0-2"
.
.
time="2020-03-10T16:24:41+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 0-9"

time="2020-03-10T16:24:41+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 1-0"
time="2020-03-10T16:24:41+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 1-1"
time="2020-03-10T16:24:42+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 1-2"
.
.
time="2020-03-10T16:24:45+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 1-9"

time="2020-03-10T16:24:45+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 2-0"
time="2020-03-10T16:24:45+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 2-1"
time="2020-03-10T16:24:46+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 2-2"
.
.
time="2020-03-10T16:24:48+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 2-9"

time="2020-03-10T16:24:49+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:24:54+05:30" level=info msg="Deleting - Bucket: cloud-nuke-test-xxxx - objects: 30"
time="2020-03-10T16:24:55+05:30" level=info msg="[OK] - 1-30 objects of batch 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:57+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:24:58+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:24:58+05:30" level=error msg="[Failed] - 1/1 - Bucket: cloud-nuke-test-xxxx - object deletion error - NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: xxxx, host id: xxxx"
--- PASS: TestNukeS3BucketWithVersioningAllObjects 
```

7. Deleting non-empty bucket - with object versioning - deletion batch size (3) < number of objects (30):

```
=== RUN   TestNukeS3BucketWithVersioningBatchObjects
time="2020-03-10T16:25:00+05:30" level=info msg="Random region chosen: us-east-2"
time="2020-03-10T16:25:00+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - creating"

time="2020-03-10T16:25:04+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 0-0"
time="2020-03-10T16:25:04+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 0-1"
time="2020-03-10T16:25:04+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 0-2"
.
.
time="2020-03-10T16:25:07+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 0-9"

time="2020-03-10T16:25:08+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 1-0"
time="2020-03-10T16:25:08+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 1-1"
time="2020-03-10T16:25:08+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 1-2"
.
.
time="2020-03-10T16:25:11+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 1-9"

time="2020-03-10T16:25:12+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f0.txt - content: 2-0"
time="2020-03-10T16:25:12+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f1.txt - content: 2-1"
time="2020-03-10T16:25:12+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f2.txt - content: 2-2"
.
.
time="2020-03-10T16:25:15+05:30" level=info msg="Bucket: cloud-nuke-test-xxxx - adding object: l1/l2/l3/f9.txt - content: 2-9"

time="2020-03-10T16:25:16+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:25:22+05:30" level=info msg="Deleting - Bucket: cloud-nuke-test-xxxx - objects: 30"

time="2020-03-10T16:25:23+05:30" level=info msg="[OK] - 1-3 objects of batch 1/10 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:25:24+05:30" level=info msg="[OK] - 4-6 objects of batch 2/10 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:25:26+05:30" level=info msg="[OK] - 7-9 objects of batch 3/10 - Bucket: cloud-nuke-test-xxxx - deleted"
.
.
time="2020-03-10T16:25:37+05:30" level=info msg="[OK] - 28-30 objects of batch 10/10 - Bucket: cloud-nuke-test-xxxx - deleted"

time="2020-03-10T16:25:39+05:30" level=info msg="[OK] - 1/1 - Bucket: cloud-nuke-test-xxxx - deleted"
time="2020-03-10T16:25:41+05:30" level=info msg="Deleting - 1 S3 Buckets in region us-east-2"
time="2020-03-10T16:25:41+05:30" level=error msg="[Failed] - 1/1 - Bucket: cloud-nuke-test-xxxx - object deletion error - NoSuchBucket: The specified bucket does not exist\n\tstatus code: 404, request id: xxxx, host id: xxxx"
--- PASS: TestNukeS3BucketWithVersioningBatchObjects
```

saurabh-hirani commented:
@brikis98 @yorinasub17  - can you please review the code and let me know if this looks good?

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,295 @@
+
+	if result.LocationConstraint == nil {
+		// GetBucketLocation returns nil for us-east-1
+		// https://github.com/aws/aws-sdk-go/issues/1687
```
Thanks for documenting this 👍

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,295 @@
+		bucketTags, err := getS3BucketTags(svc, bucket.Name)
+		if len(bucketTags) > 0 {
+			for _, tagSet := range bucketTags {
+				if tagSet["Key"] == "cloud-nuke-excluded" && tagSet["Value"] == "true" {
```
Can you add a comment here indicating why we are doing this, and include a link to my comment? (https://github.com/gruntwork-io/cloud-nuke/pull/96#pullrequestreview-367772274)

Also if you have time, adding a section in the README documenting this for discoverability would be great!

saurabh-hirani replied:
Will add comment. This is already documented in the README as ```Deleting all S3 buckets in an AWS account - except for buckets tagged with Key=cloud-nuke-excluded Value=true
```

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,295 @@
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+
+OUTER:
```
We generally prefer `return` than `continue`, especially if you have a `continue` with a label. Also, it will be easier to read if you get all the buckets in a list, and then massage it to a map as opposed to going straight to map.

For example:

```go
bucketsToInclude := []Bucket{}
for _, bucket := range output.Buckets {
  if shouldIncludeBucket(bucket) {
    bucketsToInclude = append(bucketsToInclude, bucket)
  }
}
for _, bucket := range bucketsToInclude {
  bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
  if err != nil {
    // log some message
  } else {
    // add to bucketNamesPerRegion
  }
}
```

saurabh-hirani replied:
Removed the continue OUTER - but building the map on the go as the code is readable after moving filtering code to shouldIncludeBucket.

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,295 @@
+		func(page *s3.ListObjectsV2Output, lastPage bool) (shouldContinue bool) {
+
+			for _, obj := range page.Contents {
+				if isVersioned {
```
If versioned, we should go directly to the `ListObjectVersionsPages` instead of iterating on the objects, since `ListObjectVersionsPages` without a `Prefix` will list everything in the bucket. This not only simplifies the code here, but also reduces the API calls down to one vs N, where N is the number of objects.

saurabh-hirani replied:
Done. Thanks for catching that.

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,295 @@
+				versions = append(versions, obj.VersionId)
+			}
+			return true
+		},
```
You also need to get the `DeleteMarkers` and handle those.

saurabh-hirani replied:
I might be missing something but I did not follow this - DeleteMarker is set when you do a simple Delete on a versioned bucket. But in case of a versioned bucket we are deleting each object with version so that it gets permanently deleted. https://docs.amazonaws.cn/en_us/AmazonS3/latest/dev/DeletingObjectVersions.html - because we are doing https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.DeleteObjects will we run into this?

saurabh-hirani replied:
Checked and understood. This is to handle the state when someone does a simple delete and the deletemarker is created. Will add the support.

saurabh-hirani replied:
Done

yorinasub17 commented on aws/s3_test.go:
```diff
@@ -0,0 +1,291 @@
+
+func TestNukeS3BucketWithVersioningBatchObjects(t *testing.T) {
+	testNukeS3BucketWrapper(t, true, 10, 3)
+}
```
Thanks for adding these tests! One more test request: nuking a bucket where an object was deleted on a versioned bucket. This requires using `DeleteMarkers` to empty, which you aren't handling.

saurabh-hirani replied:
Done

saurabh-hirani replied:
Will update this thread once the above query is resolved.

saurabh-hirani replied:
Done

saurabh-hirani commented:
@yorinasub17 @brikis98  - squashed the commits related to DeleteMarker, shouldIncludeBucket and single call for getting all objects and versions as per feedback. Can you please review it and let me know if it looks good?

saurabh-hirani commented:
Found a bug - if the passed in session object to ```getAllS3Buckets``` is different from bucket's target region - then the call to get bucket tags fails because it expects the S3 client to be in target region. Hence added code to maintain an internal cache of reusable per region s3 clients while collecting the buckets and updated test cases to pass in random sessions to test scenarios where default region != target bucket region. This bug was undetected till now because in the test cases we were passing in sessions which were same as bucket's target region. Also renamed  ```session``` var to ```awsSession``` to avoid confusion with the ```session``` package of the SDK.

Will squash the fix once reviewed.

yorinasub17 commented:
Thanks for working through the fixes! I am a bit buried right now to take a look, but will try to carve out time before the end of the week. Thanks for your patience!

saurabh-hirani commented:
@brikis98 @yorinasub17 - just wanted to follow up on this and wanted to know if it looks good. Btw, checked AWS landing zone - looks great.

yorinasub17 commented:
@saurabh-hirani sorry haven't had a chance to take a look yet, but this hasn't dropped off my radar. Should have time to look in the next few days.

yorinasub17 commented on aws/s3.go:
```diff
@@ -0,0 +1,306 @@
+
+	if batchSize <= 0 || batchSize > 1000 {
+		batchSize = 1000
+	}
```
I think we will want to return an error instead of silently adjusting to the max.

saurabh-hirani replied:
Done.

yorinasub17 commented on aws/s3_test.go:
```diff
@@ -0,0 +1,383 @@
+			"Value": "true",
+		},
+	})
+}
```
Can you add another test case with multiple tags, and with `cloud-nuke-excluded` not being the first element? Also, update the test function to accept a bool for determining if it should check for the bucket or not.

saurabh-hirani replied:
Done.

saurabh-hirani commented:
@yorinasub17 updated as per feedback. Please check the latest commit and if it looks good - I will squash the changes so that it can be merged.

zackproser commented on aws/s3.go:
```diff
@@ -0,0 +1,306 @@
+	bucketTags, err := getS3BucketTags(svc, bucket.Name)
+	if len(bucketTags) > 0 {
+		for _, tagSet := range bucketTags {
+			if tagSet["Key"] == "cloud-nuke-excluded" && tagSet["Value"] == "true" {
```
More of a general question / idea this gave me: It seems [tags are case sensitive](https://docs.aws.amazon.com/AmazonS3/latest/user-guide/add-object-tags.html) - perhaps we should eventually add (probably somewhere more centralized) a standard lookup feature that normalizes keys so that even if someone accidentally capitalized this tag like `Cloud-nuke-excluded` it would still have the desired effect of protecting the bucket.

saurabh-hirani replied:
Makes sense. Will convert it to lower and then compare key and value. Ideally when resources start supporting tags - the tag comparison function should be central. Right now it is local to S3.

saurabh-hirani replied:
Done as per https://github.com/gruntwork-io/cloud-nuke/pull/96/commits/9e05a356693d5d9bdb212c57ef8fc19dd3bf9909

yorinasub17 commented:
No need to squash! This looks ready to merge once the tests pass, but as mentioned this will break our environment so I needed to extend it in https://github.com/gruntwork-io/cloud-nuke/pull/101.

Note that this PR will autoclose as merged when that PR merges (because all the commits are still there).

yorinasub17 commented:
Thanks for sticking with this! This is now merged and released as https://github.com/gruntwork-io/cloud-nuke/releases/tag/v0.1.17. Binaries will show up shortly.

saurabh-hirani commented:
Thanks for the feedback and code reviews @yorinasub17 @brikis98 @zackproser!

zackproser commented:
> Thanks for the feedback and code reviews @yorinasub17 @brikis98 @zackproser!

Thank you, @saurabh-hirani ! :clap:
